source/elements/oneDAL/source/programming_model/algorithm_anatomy.rst

```diff
@@ -1,23 +1,184 @@
+
+- Aggregator of meta-parameters. Descriptor provides an interface for setting up
+  meta-parameters in compile-time or run-time.
+
+- Object that carries state of the algorithm. In the general case, descriptor is
+  stateful object whose state changes after applying an operation.
+
+Each oneDAL algorithm has its own dedicated namespace, where corresponding
+descriptor is defined. Descriptor, in its turn, defines a set of template
+parameters and **properties** --- run-time parameters that can be accessed by
+means of the corresponding getter and setter methods. `The following code sample
+<descriptor-template_>`_ shows the common structure of descriptor's definition
+for an abstract algorithm. To define a particular algorithm the following
+strings shall be substituted:
+
+- ``%ALGORITHM%`` is the name of algorithm and its namespace. All classes and
+  structures related to that algorithm shall be defined within the namespace.
+
+- ``%PROPERTY_NAME%`` and ``%PROPERTY_TYPE%`` are the name and type of the one
+  of algorithm's properties.
+
+.. code-block:: cpp
+   :name: descriptor-template
+
+   namespace onedal::%ALGORITHM% {
+
+   template <typename Float  = default_float_t,
+             typename Method = method::by_default,
+             /* more template parameters */>
+   class descriptor {
+   public:
+      /* Getter & Setter for the property called `%PROPERTY_NAME%` */
+      descriptor& set_%PROPERTY_NAME%(%PROPERTY_TYPE% value);
+      %PROPERTY_TYPE% get_%PROPERTY_NAME%() const;
+
+      /* more properties */
+   };
+
+   } // namespace onedal::%ALGORITHM%
+
+
+Each algorithm's meta-parameter is mapped to the property that shall satisfy the
+following requirements:
+
+- Properties are defined by means of getter and setter method. The underlying
+  class field that stores the property's value is never exposed in the
+  descriptor interface.
+
+- The getter returns value of the underlying class field.
+
+- The setter accepts the only parameter of property's type and assigns it
+  to the underlying class field.
+
+- Most of the properties are preset with default values, but some of them
+  are initialized by passing required parameters to the constructor.
+
+- The setter returns a reference to the descriptor object to allow chaining
+  calls as shown in the example below.
+
+  .. code-block:: cpp
+
+     auto desc = descriptor{}
+        .set_property_name_1(value_1)
+        .set_property_name_2(value_2)
+        .set_property_name_3(value_3);
+
+
+Descriptor is allowed to have any number of template parameters, but shall
+support at least two:
+
+- ``Float`` is a `floating-point type <floating-point_>`_ that the algorithm
+  uses for computations. This parameter is defined first and has the
+  ``onedal::default_float_t`` default value.
+
+- ``Method`` is a tag-type that specifies an `computational method <methods_>`_.
+  This parameter is defined second and has the ``method::by_default`` default
+  value.
```

Comment:
I think those lines are logically belong to the part where template parameters are first introduced